### ISSUE TITLE
'border' and 'reflection’ modes of grid_sample have incorrect gradients at border

### ISSUE THREAD

**Issue by bnehoran** [pytorch/pytorch#23925]
## 🐛 Bug

When `padding_mode='border'` in `grid_sample`, and a grid point falls exactly on the high boundary of the image (`size - 1`), the gradient should be based on the border padding scheme, which should give either the gradient from just inside the boundary, or zero from just outside the boundary (either could be valid, since it’s a non differentiable point). Instead, the gradient is currently based on zero padding the image, which gives wacky results.

Same problem occurs with `padding_mode='reflection'` for 2D `grid_sample` on CPU.
Reflection modes of both the cuda version and the 3D CPU version also have this problem, but it’s arguably worse, since the incorrect gradient is also negated. Furthermore, this is an inconsistency between the behavior of CPU and CUDA kernels.

## Example:
```python
image = torch.arange(0, 5, dtype=torch.float).expand((1,1,5,5)).requires_grad_()

id_grid = torch.nn.functional.affine_grid(
    torch.tensor([[[1,0,0],[0,1,0.]]]), (1,1,5,5), align_corners=True).requires_grad_()

torch.nn.functional.grid_sample(image, id_grid, padding_mode='border',
                                align_corners=True).sum().backward()

print(id_grid.grad.permute(0,3,1,2))
```
```python
tensor([[[[ 2.,  2.,  2.,  2., -8.],
          [ 2.,  2.,  2.,  2., -8.],
          [ 2.,  2.,  2.,  2., -8.],
          [ 2.,  2.,  2.,  2., -8.],
          [ 2.,  2.,  2.,  2., -8.]],

         [[ 0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.],
          [ 0., -2., -4., -6., -8.]]]])
```

Notice the wacky last row and last column. This is because the gradient there is currently calculated as if the image was zero-padded.

The result should ideally look like

```python
tensor([[[[ 2.,  2.,  2.,  2.,  2.],
          [ 2.,  2.,  2.,  2.,  2.],
          [ 2.,  2.,  2.,  2.,  2.],
          [ 2.,  2.,  2.,  2.,  2.],
          [ 2.,  2.,  2.,  2.,  2.]],

         [[ 0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.]]]])
```

which finds the gradient using the in-bounds neighbor.

A less ideal, but still palatable result would be

```python
tensor([[[[ 2.,  2.,  2.,  2.,  0.],
          [ 2.,  2.,  2.,  2.,  0.],
          [ 2.,  2.,  2.,  2.,  0.],
          [ 2.,  2.,  2.,  2.,  0.],
          [ 2.,  2.,  2.,  2.,  0.]],

         [[ 0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.]]]])
```

which finds the gradient using the out-of-bounds, border-padded neighbor.

Reflection mode on cpu (for instance, try using these same commands, but with `padding_mode='reflection'`) gives the exact same problematic result.
When using reflection mode on cuda, however, (as well as for 3D grid_sample on cpu) the problematic gradients are negated!

```python
tensor([[[[2., 2., 2., 2., 8.],
          [2., 2., 2., 2., 8.],
          [2., 2., 2., 2., 8.],
          [2., 2., 2., 2., 8.],
          [2., 2., 2., 2., 8.]],

         [[0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.],
          [-0., 2., 4., 6., 8.]]]])
```

This is also problematic, of course, but even more so because of the mismatch between the cpu and cuda behaviors.

For `reflection` mode, I think it makes sense to set the gradient in such cases to zero, since it’s sort of at the apex of a symmetric hill. But setting it to take the gradient of one side or the other might also be acceptable for most practical purposes.

For `border` mode, by contrast, I think it makes more sense to always take the non-zero gradient from the inner side, since the outer side gradient will be zero and so effectively stop training (see the related discussion for `clamp` at #7002 and #7049).

PyTorch Version: tested on commit https://github.com/pytorch/pytorch/commit/0539462ca2966aa29657b58aeb17a85c21524d31

**Comment by Eurus-Holmes** [comment]
I have also the same confusion, in `grid_sample` source implementation:

https://github.com/pytorch/pytorch/blob/f064c5aa33483061a48994608d890b968ae53fb5/aten/src/THNN/generic/SpatialGridSamplerBilinear.c

I don't find the implementation of 'reflection' padding mode, I even don't understand the `padding mode` how to implement, somebody could help me?

**Comment by bnehoran** [comment]
@Eurus-Holmes what you're linking to is an old, outdated version of the code.

If you're having trouble using `grid_sample` and would like help with that, feel free to post on the PyTorch forum at discuss.pytorch.org, and we'll do our best to help you out there.

**Comment by bnehoran** [comment]
This has a real effect on training, since any grid points that are supposed to end up at the border of the image are instead sent off with a wild gradient in some undesired direction.

Here's an example result from a convnet trained to do image registration. It predicts a dense displacement field, and passes that to `grid_sample` to warp one input image to another. Here it predicts no displacement between the two input images, but take a look at what the grid points that should be at the border are doing instead:
![image](https://user-images.githubusercontent.com/9757500/63630142-b49ecf00-c5e5-11e9-96b9-008061aacc04.png)
Each of the vectors in the figure represents the displacement of the corresponding grid point away from the identity grid. This convnet uses `grid_sample` with `bilinear` interpolation mode and `border` padding mode.

Note that this problem is not likely to be caught by `gradcheck` because the incorrect gradient only occurs when a grid point happens to fall exactly on the border, which is not likely to happen by accident.

As you can see in this example, however, when a convnet is in fact trying to predict no displacement, the border grid points will eventually fall exactly at the border, and then it will be thrown off by updating with the wrong gradient. So it just never settles.

I posted my suggestions for what the gradient at the borders should look like up above, but the fix would require a bit of thought in order to implement cleanly, because in order to get the gradient on the inner side for `border` padding mode, there might have to be a special case for grid points at the border to shift over by 1 the four pixels (or eight voxels in the 3D case) that it uses for the bilinear interpolation.

@fmassa and @SsnL what do you think?

**Comment by bnehoran** [comment]
I've pushed a fix for this issue at https://github.com/pytorch/pytorch/pull/32829
